File: Aether Mobile SDK/Aether.swift
// =============================================================================
// AETHER SDK — iOS (Swift)
// Core analytics, identity, session, consent, Web3 tracking
// =============================================================================

import Foundation
import UIKit

// MARK: - Configuration

public struct AetherConfig {
    public let apiKey: String
    public var environment: Environment = .production
    public var debug: Bool = false
    public var endpoint: String = "https://api.aether.network"
    public var modules: ModuleConfig = ModuleConfig()
    public var privacy: PrivacyConfig = PrivacyConfig()
    public var batchSize: Int = 10
    public var flushInterval: TimeInterval = 5.0
    
    public init(apiKey: String) {
        self.apiKey = apiKey
    }
    
    public enum Environment: String, Codable {
        case production, staging, development
    }
}

public struct ModuleConfig {
    public var screenTracking: Bool = true
    public var deepLinkAttribution: Bool = true
    public var pushNotificationTracking: Bool = true
    public var walletTracking: Bool = false
    public var purchaseTracking: Bool = true
    public var errorTracking: Bool = true
    public var experiments: Bool = true
}

public struct PrivacyConfig {
    public var gdprMode: Bool = false
    public var anonymizeIP: Bool = true
    public var respectATT: Bool = true
}

// MARK: - Event Types

public enum AetherEventType: String, Codable {
    case track, screen, identify, conversion, wallet, transaction, error, consent
}

public struct AetherEvent: Codable {
    public let id: String
    public let type: AetherEventType
    public let timestamp: String
    public let sessionId: String
    public let anonymousId: String
    public var userId: String?
    public var properties: [String: AnyCodable]
    public var context: EventContext
}

public struct EventContext: Codable {
    public let library: LibraryInfo
    public var device: DeviceInfo?
    public var campaign: CampaignInfo?
    
    public struct LibraryInfo: Codable {
        public let name: String
        public let version: String
    }
    
    public struct DeviceInfo: Codable {
        public let type: String
        public let os: String
        public let osVersion: String
        public let model: String
        public let manufacturer: String
        public let screenWidth: Int
        public let screenHeight: Int
        public let language: String
        public let timezone: String
    }
    
    public struct CampaignInfo: Codable {
        public var source: String?
        public var medium: String?
        public var campaign: String?
    }
}

// MARK: - Identity

public struct IdentityData {
    public var userId: String?
    public var walletAddress: String?
    public var walletType: String?
    public var chainId: Int?
    public var traits: [String: AnyCodable]?
    
    public init(userId: String? = nil, walletAddress: String? = nil, traits: [String: AnyCodable]? = nil) {
        self.userId = userId
        self.walletAddress = walletAddress
        self.traits = traits
    }
}

// MARK: - Main SDK Class

public final class Aether {
    public static let shared = Aether()
    
    private var config: AetherConfig?
    private var eventQueue: [AetherEvent] = []
    private var sessionId: String = UUID().uuidString
    private var anonymousId: String = ""
    private var userId: String?
    private var traits: [String: AnyCodable] = [:]
    private var flushTimer: Timer?
    private var sessionStart: Date = Date()
    private var screenCount: Int = 0
    private var eventCount: Int = 0
    private var isInitialized = false
    
    private let serialQueue = DispatchQueue(label: "com.aether.sdk.serial")
    private let defaults = UserDefaults(suiteName: "com.aether.sdk")!
    
    private init() {}
    
    // MARK: - Public API
    
    public func initialize(config: AetherConfig) {
        guard !isInitialized else {
            log("Already initialized")
            return
        }
        
        self.config = config
        self.anonymousId = loadOrCreateAnonymousId()
        self.sessionId = UUID().uuidString
        self.sessionStart = Date()
        
        // Setup flush timer
        flushTimer = Timer.scheduledTimer(withTimeInterval: config.flushInterval, repeats: true) { [weak self] _ in
            self?.flush()
        }
        
        // Setup lifecycle observers
        setupLifecycleObservers()
        
        // Auto screen tracking via swizzling
        if config.modules.screenTracking {
            UIViewController.swizzleViewDidAppear()
        }
        
        isInitialized = true
        log("Aether iOS SDK initialized (v4.0.0)")
    }
    
    public func track(_ event: String, properties: [String: AnyCodable] = [:]) {
        enqueueEvent(type: .track, properties: ["event": AnyCodable(event)].merging(properties) { _, new in new })
    }
    
    public func screenView(_ screenName: String, properties: [String: AnyCodable] = [:]) {
        screenCount += 1
        enqueueEvent(type: .screen, properties: ["screen": AnyCodable(screenName)].merging(properties) { _, new in new })
    }
    
    public func conversion(_ event: String, value: Double? = nil, properties: [String: AnyCodable] = [:]) {
        var props = properties
        props["event"] = AnyCodable(event)
        if let value = value { props["value"] = AnyCodable(value) }
        enqueueEvent(type: .conversion, properties: props)
    }
    
    public func hydrateIdentity(_ data: IdentityData) {
        if let userId = data.userId { self.userId = userId }
        if let traits = data.traits { self.traits.merge(traits) { _, new in new } }
        
        enqueueEvent(type: .identify, properties: [
            "userId": AnyCodable(userId ?? ""),
            "traits": AnyCodable(traits),
            "walletAddress": AnyCodable(data.walletAddress ?? "")
        ])
        
        // Persist
        defaults.set(userId, forKey: "userId")
    }
    
    public func getAnonymousId() -> String { anonymousId }
    public func getUserId() -> String? { userId }
    
    public func reset() {
        flush()
        userId = nil
        traits = [:]
        anonymousId = UUID().uuidString
        sessionId = UUID().uuidString
        defaults.removeObject(forKey: "userId")
        defaults.set(anonymousId, forKey: "anonymousId")
        log("SDK reset")
    }
    
    public func flush() {
        serialQueue.async { [weak self] in
            self?.sendBatch()
        }
    }
    
    // MARK: - Deep Link Attribution
    
    public func handleDeepLink(_ url: URL) {
        let components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        var attribution: [String: AnyCodable] = ["url": AnyCodable(url.absoluteString)]
        
        for item in components?.queryItems ?? [] {
            if item.name.hasPrefix("utm_") || item.name == "gclid" || item.name == "fbclid" {
                attribution[item.name] = AnyCodable(item.value ?? "")
            }
        }
        
        track("deep_link_opened", properties: attribution)
    }
    
    // MARK: - Push Notification
    
    public func trackPushOpened(userInfo: [AnyHashable: Any]) {
        var props: [String: AnyCodable] = [:]
        if let campaignId = userInfo["campaign_id"] as? String {
            props["campaignId"] = AnyCodable(campaignId)
        }
        track("push_notification_opened", properties: props)
    }
    
    // MARK: - Private
    
    private func enqueueEvent(type: AetherEventType, properties: [String: AnyCodable]) {
        guard isInitialized else { return }
        
        let event = AetherEvent(
            id: UUID().uuidString,
            type: type,
            timestamp: ISO8601DateFormatter().string(from: Date()),
            sessionId: sessionId,
            anonymousId: anonymousId,
            userId: userId,
            properties: properties,
            context: buildContext()
        )
        
        serialQueue.async { [weak self] in
            self?.eventQueue.append(event)
            if let batchSize = self?.config?.batchSize, (self?.eventQueue.count ?? 0) >= batchSize {
                self?.sendBatch()
            }
        }
        
        eventCount += 1
    }
    
    private func sendBatch() {
        guard !eventQueue.isEmpty, let config = config else { return }
        
        let batch = Array(eventQueue.prefix(config.batchSize))
        eventQueue.removeFirst(min(batch.count, eventQueue.count))
        
        guard let url = URL(string: "\(config.endpoint)/v1/batch") else { return }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(config.apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("ios", forHTTPHeaderField: "X-Aether-SDK")
        
        let payload: [String: Any] = [
            "batch": batch.map { try? JSONEncoder().encode($0) }.compactMap { $0 }.map { try? JSONSerialization.jsonObject(with: $0) }.compactMap { $0 },
            "sentAt": ISO8601DateFormatter().string(from: Date())
        ]
        
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload)
        
        URLSession.shared.dataTask(with: request) { [weak self] _, response, error in
            if let error = error {
                self?.log("Batch send failed: \(error.localizedDescription)")
                // Re-enqueue failed events
                self?.serialQueue.async {
                    self?.eventQueue.insert(contentsOf: batch, at: 0)
                }
            } else if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode >= 400 {
                self?.log("Batch send HTTP error: \(httpResponse.statusCode)")
            }
        }.resume()
    }
    
    private func buildContext() -> EventContext {
        let device = UIDevice.current
        let screen = UIScreen.main
        
        return EventContext(
            library: .init(name: "AetherSDK-iOS", version: "4.0.0"),
            device: .init(
                type: device.userInterfaceIdiom == .pad ? "tablet" : "mobile",
                os: "iOS",
                osVersion: device.systemVersion,
                model: device.model,
                manufacturer: "Apple",
                screenWidth: Int(screen.bounds.width * screen.scale),
                screenHeight: Int(screen.bounds.height * screen.scale),
                language: Locale.current.language.languageCode?.identifier ?? "en",
                timezone: TimeZone.current.identifier
            )
        )
    }
    
    private func loadOrCreateAnonymousId() -> String {
        if let stored = defaults.string(forKey: "anonymousId") {
            return stored
        }
        let id = UUID().uuidString
        defaults.set(id, forKey: "anonymousId")
        return id
    }
    
    private func setupLifecycleObservers() {
        NotificationCenter.default.addObserver(forName: UIApplication.didEnterBackgroundNotification, object: nil, queue: .main) { [weak self] _ in
            self?.flush()
        }
        NotificationCenter.default.addObserver(forName: UIApplication.willTerminateNotification, object: nil, queue: .main) { [weak self] _ in
            self?.flush()
        }
        NotificationCenter.default.addObserver(forName: UIApplication.willEnterForegroundNotification, object: nil, queue: .main) { [weak self] _ in
            self?.sessionId = UUID().uuidString
            self?.sessionStart = Date()
            self?.track("app_foreground")
        }
    }
    
    private func log(_ message: String) {
        guard config?.debug == true else { return }
        print("[Aether SDK] \(message)")
    }
}

// MARK: - UIViewController Swizzling for Auto Screen Tracking

extension UIViewController {
    static var hasSwizzled = false
    
    static func swizzleViewDidAppear() {
        guard !hasSwizzled else { return }
        hasSwizzled = true
        
        let originalSelector = #selector(UIViewController.viewDidAppear(_:))
        let swizzledSelector = #selector(UIViewController.aether_viewDidAppear(_:))
        
        guard let originalMethod = class_getInstanceMethod(UIViewController.self, originalSelector),
              let swizzledMethod = class_getInstanceMethod(UIViewController.self, swizzledSelector) else { return }
        
        method_exchangeImplementations(originalMethod, swizzledMethod)
    }
    
    @objc func aether_viewDidAppear(_ animated: Bool) {
        aether_viewDidAppear(animated) // Calls original
        
        let screenName = String(describing: type(of: self))
        let ignoredPrefixes = ["UI", "_", "NS"]
        if !ignoredPrefixes.contains(where: { screenName.hasPrefix($0) }) {
            Aether.shared.screenView(screenName)
        }
    }
}

// MARK: - AnyCodable Helper

public struct AnyCodable: Codable {
    public let value: Any
    
    public init(_ value: Any) { self.value = value }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let v = try? container.decode(String.self) { value = v }
        else if let v = try? container.decode(Int.self) { value = v }
        else if let v = try? container.decode(Double.self) { value = v }
        else if let v = try? container.decode(Bool.self) { value = v }
        else if let v = try? container.decode([String: AnyCodable].self) { value = v }
        else if let v = try? container.decode([AnyCodable].self) { value = v }
        else { value = "" }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch value {
        case let v as String: try container.encode(v)
        case let v as Int: try container.encode(v)
        case let v as Double: try container.encode(v)
        case let v as Bool: try container.encode(v)
        case let v as [String: AnyCodable]: try container.encode(v)
        case let v as [AnyCodable]: try container.encode(v)
        default: try container.encode(String(describing: value))
        }
    }
}

File: Aether Mobile SDK/Aether.kt
// =============================================================================
// AETHER SDK — Android (Kotlin)
// Core analytics, identity, session, consent, lifecycle tracking
// =============================================================================

package com.aether.sdk

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import android.util.DisplayMetrics
import android.util.Log
import androidx.lifecycle.*
import kotlinx.coroutines.*
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

// =============================================================================
// CONFIGURATION
// =============================================================================

data class AetherConfig(
    val apiKey: String,
    val environment: Environment = Environment.PRODUCTION,
    val debug: Boolean = false,
    val endpoint: String = "https://api.aether.network",
    val batchSize: Int = 10,
    val flushIntervalMs: Long = 5000L,
    val modules: ModuleConfig = ModuleConfig(),
    val privacy: PrivacyConfig = PrivacyConfig()
) {
    enum class Environment { PRODUCTION, STAGING, DEVELOPMENT }
}

data class ModuleConfig(
    val activityTracking: Boolean = true,
    val deepLinkAttribution: Boolean = true,
    val pushTracking: Boolean = true,
    val walletTracking: Boolean = false,
    val purchaseTracking: Boolean = true,
    val errorTracking: Boolean = true,
    val experiments: Boolean = true
)

data class PrivacyConfig(
    val gdprMode: Boolean = false,
    val anonymizeIP: Boolean = true
)

// =============================================================================
// IDENTITY
// =============================================================================

data class IdentityData(
    val userId: String? = null,
    val walletAddress: String? = null,
    val walletType: String? = null,
    val chainId: Int? = null,
    val traits: Map<String, Any?> = emptyMap()
)

// =============================================================================
// MAIN SDK
// =============================================================================

object Aether : DefaultLifecycleObserver {
    private const val TAG = "AetherSDK"
    private const val VERSION = "5.0.0"
    private const val PREFS_NAME = "com.aether.sdk"

    private var config: AetherConfig? = null
    private var context: Context? = null
    private var prefs: SharedPreferences? = null

    private val eventQueue = ConcurrentLinkedQueue<JSONObject>()
    private var flushJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private var sessionId: String = UUID.randomUUID().toString()
    private var anonymousId: String = ""
    private var userId: String? = null
    private var traits: MutableMap<String, Any?> = mutableMapOf()
    private var screenCount = 0
    private var eventCount = 0
    private var isInitialized = false
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }

    // =========================================================================
    // PUBLIC API
    // =========================================================================

    fun initialize(application: Application, config: AetherConfig) {
        if (isInitialized) {
            log("Already initialized")
            return
        }

        this.config = config
        this.context = application.applicationContext
        this.prefs = application.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        this.anonymousId = loadOrCreateAnonymousId()
        this.userId = prefs?.getString("userId", null)
        this.sessionId = UUID.randomUUID().toString()

        // Lifecycle tracking
        ProcessLifecycleOwner.get().lifecycle.addObserver(this)

        // Auto Activity tracking
        if (config.modules.activityTracking) {
            application.registerActivityLifecycleCallbacks(ActivityTracker())
        }

        // Uncaught exception handler
        if (config.modules.errorTracking) {
            setupErrorTracking()
        }

        // Start flush timer
        startFlushTimer()

        // Initialize semantic context collector
        SemanticContextCollector.initialize(application.applicationContext)
        SemanticContextCollector.resetSession()

        // Initialize Web2 modules
        AetherEcommerce.initialize(
            context = application.applicationContext,
            apiKey = config.apiKey,
            endpoint = config.endpoint
        )
        AetherFeatureFlags.initialize(
            context = application.applicationContext,
            apiKey = config.apiKey,
            endpoint = config.endpoint
        )
        AetherFeedback.initialize(
            context = application.applicationContext,
            apiKey = config.apiKey,
            endpoint = config.endpoint,
            userId = anonymousId
        )

        isInitialized = true
        log("Aether Android SDK initialized (v$VERSION) — Web2 + Web3 modules enabled")

        // Start OTA data module update manager (non-blocking, background)
        AetherUpdateManager.start(
            context = application.applicationContext,
            apiKey = config.apiKey,
            endpoint = config.endpoint,
            currentVersion = VERSION
        )
    }

    fun track(event: String, properties: Map<String, Any?> = emptyMap()) {
        val props = mutableMapOf<String, Any?>("event" to event)
        props.putAll(properties)
        enqueueEvent("track", props)
    }

    fun screenView(screenName: String, properties: Map<String, Any?> = emptyMap()) {
        screenCount++
        SemanticContextCollector.recordScreen(screenName)
        val props = mutableMapOf<String, Any?>("screen" to screenName)
        props.putAll(properties)
        enqueueEvent("screen", props)
    }

    fun conversion(event: String, value: Double? = null, properties: Map<String, Any?> = emptyMap()) {
        val props = mutableMapOf<String, Any?>("event" to event)
        if (value != null) props["value"] = value
        props.putAll(properties)
        enqueueEvent("conversion", props)
    }

    fun hydrateIdentity(data: IdentityData) {
        data.userId?.let { userId = it }
        traits.putAll(data.traits)

        val props = mutableMapOf<String, Any?>(
            "userId" to (userId ?: ""),
            "traits" to traits,
            "walletAddress" to (data.walletAddress ?: "")
        )
        enqueueEvent("identify", props)

        prefs?.edit()?.putString("userId", userId)?.apply()
    }

    fun getAnonymousId(): String = anonymousId
    fun getUserId(): String? = userId

    fun reset() {
        flush()
        userId = null
        traits.clear()
        anonymousId = UUID.randomUUID().toString()
        sessionId = UUID.randomUUID().toString()
        prefs?.edit()
            ?.remove("userId")
            ?.putString("anonymousId", anonymousId)
            ?.apply()
        log("SDK reset")
    }

    fun flush() {
        scope.launch { sendBatch() }
    }

    fun handleDeepLink(url: String) {
        try {
            val uri = java.net.URI(url)
            val params = uri.query?.split("&")?.associate {
                val parts = it.split("=", limit = 2)
                parts[0] to (parts.getOrNull(1) ?: "")
            } ?: emptyMap()

            val attribution = mutableMapOf<String, Any?>("url" to url)
            params.forEach { (key, value) ->
                if (key.startsWith("utm_") || key in listOf("gclid", "fbclid", "msclkid")) {
                    attribution[key] = value
                }
            }
            track("deep_link_opened", attribution)
        } catch (e: Exception) {
            log("Failed to parse deep link: ${e.message}")
        }
    }

    fun trackPushOpened(data: Map<String, String>) {
        track("push_notification_opened", mapOf(
            "campaignId" to (data["campaign_id"] ?: ""),
            "messageId" to (data["message_id"] ?: "")
        ))
    }

    // =========================================================================
    // LIFECYCLE
    // =========================================================================

    override fun onStart(owner: LifecycleOwner) {
        sessionId = UUID.randomUUID().toString()
        track("app_foreground")
    }

    override fun onStop(owner: LifecycleOwner) {
        track("app_background")
        flush()
    }

    // =========================================================================
    // PRIVATE
    // =========================================================================

    private fun enqueueEvent(type: String, properties: Map<String, Any?>) {
        if (!isInitialized) return

        // Inject tiered semantic context into every event
        val enrichedProperties = properties.toMutableMap()
        try {
            enrichedProperties["_semantic"] = SemanticContextCollector.collect()
        } catch (_: Exception) {}

        val event = JSONObject().apply {
            put("id", UUID.randomUUID().toString())
            put("type", type)
            put("timestamp", dateFormat.format(Date()))
            put("sessionId", sessionId)
            put("anonymousId", anonymousId)
            put("userId", userId ?: JSONObject.NULL)
            put("properties", JSONObject(enrichedProperties.mapValues { it.value ?: JSONObject.NULL }))
            put("context", buildContext())
        }

        eventQueue.add(event)
        eventCount++

        if (eventQueue.size >= (config?.batchSize ?: 10)) {
            scope.launch { sendBatch() }
        }
    }

    private suspend fun sendBatch() = withContext(Dispatchers.IO) {
        val cfg = config ?: return@withContext
        if (eventQueue.isEmpty()) return@withContext

        val batch = mutableListOf<JSONObject>()
        repeat(minOf(cfg.batchSize, eventQueue.size)) {
            eventQueue.poll()?.let { batch.add(it) }
        }
        if (batch.isEmpty()) return@withContext

        try {
            val url = URL("${cfg.endpoint}/v1/batch")
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.setRequestProperty("Authorization", "Bearer ${cfg.apiKey}")
            connection.setRequestProperty("X-Aether-SDK", "android")
            connection.doOutput = true
            connection.connectTimeout = 10000
            connection.readTimeout = 10000

            val payload = JSONObject().apply {
                put("batch", JSONArray(batch))
                put("sentAt", dateFormat.format(Date()))
            }

            connection.outputStream.use { it.write(payload.toString().toByteArray()) }

            val responseCode = connection.responseCode
            if (responseCode >= 400) {
                log("Batch send HTTP error: $responseCode")
                batch.forEach { eventQueue.add(it) } // Re-enqueue
            }
            connection.disconnect()
        } catch (e: Exception) {
            log("Batch send failed: ${e.message}")
            batch.forEach { eventQueue.add(it) }
        }
    }

    private fun buildContext(): JSONObject {
        val ctx = context ?: return JSONObject()
        val dm = ctx.resources.displayMetrics

        return JSONObject().apply {
            put("library", JSONObject().apply {
                put("name", "AetherSDK-Android")
                put("version", VERSION)
            })
            put("device", JSONObject().apply {
                put("type", if (ctx.resources.configuration.smallestScreenWidthDp >= 600) "tablet" else "mobile")
                put("os", "Android")
                put("osVersion", Build.VERSION.RELEASE)
                put("model", Build.MODEL)
                put("manufacturer", Build.MANUFACTURER)
                put("screenWidth", dm.widthPixels)
                put("screenHeight", dm.heightPixels)
                put("density", dm.density)
                put("language", Locale.getDefault().language)
                put("timezone", TimeZone.getDefault().id)
            })
        }
    }

    private fun loadOrCreateAnonymousId(): String {
        prefs?.getString("anonymousId", null)?.let { return it }
        val id = UUID.randomUUID().toString()
        prefs?.edit()?.putString("anonymousId", id)?.apply()
        return id
    }

    private fun startFlushTimer() {
        flushJob?.cancel()
        flushJob = scope.launch {
            while (isActive) {
                delay(config?.flushIntervalMs ?: 5000L)
                sendBatch()
            }
        }
    }

    private fun setupErrorTracking() {
        val defaultHandler = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            try {
                enqueueEvent("error", mapOf(
                    "type" to "uncaught_exception",
                    "message" to (throwable.message ?: "Unknown"),
                    "stack" to (throwable.stackTraceToString().take(2000)),
                    "thread" to thread.name
                ))
                runBlocking { sendBatch() }
            } catch (_: Exception) {}
            defaultHandler?.uncaughtException(thread, throwable)
        }
    }

    private fun log(message: String) {
        if (config?.debug == true) Log.d(TAG, message)
    }

    // =========================================================================
    // ACTIVITY TRACKER
    // =========================================================================

    private class ActivityTracker : Application.ActivityLifecycleCallbacks {
        override fun onActivityResumed(activity: android.app.Activity) {
            val name = activity.javaClass.simpleName
            if (!name.startsWith("_")) {
                screenView(name)
            }
        }
        override fun onActivityCreated(a: android.app.Activity, s: android.os.Bundle?) {}
        override fun onActivityStarted(a: android.app.Activity) {}
        override fun onActivityPaused(a: android.app.Activity) {}
        override fun onActivityStopped(a: android.app.Activity) {}
        override fun onActivitySaveInstanceState(a: android.app.Activity, s: android.os.Bundle) {}
        override fun onActivityDestroyed(a: android.app.Activity) {}
    }
}
